### DISCUSSION TITLE
Delay between posting and API availability

### DISCUSSION THREAD

**Discussion by GreenFootballs** [bluesky-social/atproto#1562] [Dev Debugging] [answered]
I'm working on a client, and one of the things I need to do is save a copy of the post data immediately after posting. I was doing this by calling `com.atproto.repo.createRecord`, then passing the returned uri to `app.bsky.feed.getPosts`. But in some cases (not always!), it seems like the post data is not available through the API that quickly, and `getPosts` doesn't return an error -- it's just a null array. If I wait a little while, then try the `getPosts` call again, it works as expected.

Is there some kind of caching that would cause this delay? And is there any way to work around this?

**Comment by snarfed** [comment]
Hi! This is probably because of the [PDS/BGS/AppView architecture](https://blueskyweb.xyz/blog/5-5-2023-federation-architecture).

Originally, the PDS handled all of those API calls itself, so they were strongly consistent more or less automatically. The team is now migrating toward serving reads from the AppView instead, either directly and/or proxied through the PDS. Commits take time to propagate from the PDS through the BGS to the AppView via `subscribeRepos`, so that's probably causing the eventual consistency you're seeing.

(I don't know what the team thinks of this eventually consistency. I'm curious to hear!)

**Reply by bnewbold** [reply]
We have thought about the eventual consistency a whole lot!

The situation we find the most important and time sensitive is that folks "read their own writes", meaning that if they take any action (create a record), then immediately do a fetch from the appview, it should be clear if the action is included, and best-case the action is represented. Many apps handle this in clients using a local cache, but we are trying to move this responsibility to the PDS.

One helper for this is to have the AppView return the most recently indexed repo rev value (logical clock, as a TID) in an HTTP header. These requests are proxied by the PDS, and the PDS always knows the current actual rev clock for the repo (the end app client probably also knows, but in a multi-device situation it could be out of date). So the PDS (and probably app) know if it got a stale read from the AppView or not.

One thing an application-aware PDS could do in this situation is inject in known local info to the response before returning it to the end client. For example, if you liked a post very recently, and the appview hasn't indexed that, then the "my account state" in a feed response might not include the correct "have I liked this post" metadata. The PDS "knows" that you recently liked it and can update the response.

We do this currently in our PDS reference implementation, with some limits around which responses can be mutated/updated, and IIRC how long of a "staleness" window will be handled. I can't remember the exact details, we should probably do a dev blog post about this and update the specs with the HTTP header name and semantics.

**Comment by GreenFootballs** [comment]
A little while ago I suggested that the `createRecord` call should return the entire post record, instead of just a DID and CID. I still think that would be a good practice, and it would be consistent with other social media systems -- both Mastodon and Twitter return the full post data on creation.

**Reply by snarfed** [reply, marked as answer]
Definitely! Makes sense.

The other thing you might try is `com.atproto.repo.getRecord`, which ideally the PDS can/should handle itself, instead of `app.bsky.feed.getPosts`, which is the AppView's domain.

**Reply by GreenFootballs** [reply]
Unfortunately, `getRecord` doesn't include image URLs.

**Reply by DavidBuchanan314** [reply]
That's because they're not part of the record, they're part of the view (i.e. generated and returned by the AppView, not the PDS). Image URLs are deterministic though, so you ought to be able to predict them. I'm not sure if the "format" is documented anywhere, but you can inspect existing images to figure it out.

**Reply by DavidBuchanan314** [reply]
I'm not sure I understand the purpose having the record reflected back upon creation, it ought to be exactly the same as what was submitted?

**Reply by GreenFootballs** [reply]
I thought the post was created with the image URLs already embedded, so they'd be part of the returned data. But now I see they're actually created in the AppView, so, uh, never mind!

**Reply by GreenFootballs** [reply]
`getRecord` doesn't return the _exact_ data that was submitted, though; the structure's similar but with some differences.

**Reply by DavidBuchanan314** [reply]
What's different? It ought to be truly identical, at least, semantically (object field re-ordering may occur). There's also the exception where the PDS might inject the `$type` field for you, if you omitted it (which imho it should not do).

**Reply by GreenFootballs** [reply]
The object has `cid` and `uri` keys, and the submitted data is stored under a `value` key, so it's slightly different.

**Reply by DavidBuchanan314** [reply]
The content of the `value` key *is* the record. The cid is derived from the value of the record, and the URI is knowable based on the response from createRecord (the cid is there, too)

**Comment by GreenFootballs** [comment]
P.S. For anyone following this discussion, I just discovered that the URL format for images has changed; it's now possible to get the image URLs with just the CID and the DID of the user. The image URL formats for thumb and full size:

```
https://av-cdn.bsky.app/img/feed_thumbnail/plain/__DID__/__CID__@jpeg
https://av-cdn.bsky.app/img/feed_fullsize/plain/__DID__/__CID__@jpeg
```

**Reply by GreenFootballs** [reply]
And on this note, it would be great if you could get the original image (at its original size unscaled) by using something like:

`https://av-cdn.bsky.app/img/feed_original/plain/__DID__/__CID__@jpeg`

**Reply by DavidBuchanan314** [reply]
you can get the original image via getBlob (but maybe there could be some performance benefits to getting it via a CDN)

**Reply by GreenFootballs** [reply]
The problem with getBlob is that you get... a binary blob. Not as easy to deal with as an image URL, but yes, it can work with some more finagling.

**Reply by snarfed** [reply]
`getBlob` HTTP responses include `Content-Type`!...so you can use them in `a href`s etc directly. Example: https://bsky.social/xrpc/com.atproto.sync.getBlob?did=did:plc:vwzwgnygau7ed7b7wt5ux7y2&cid=bafkreih2xfvvjo3xfvbidesyglzhip7szi6f4bbb7ocw5zrturl7wpynx4 , response includes `Content-Type: image/svg+xml`.

**Reply by GreenFootballs** [reply]
Good idea.

**Comment by GreenFootballs** [comment]
Interesting that getBlob is one of the few magic `GET` endpoints that doesn't require authentication. Is that deliberate?

**Reply by bnewbold** [reply]
Like getRepo and the firehose, this is necessary for federation and redistribution of content.
